sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,569 @@
+  def estimate: Option[Statistics] = {
+    // We first copy child node's statistics and then modify it based on filter selectivity.
+    val stats: Statistics = plan.child.stats(catalystConf)
+    if (stats.rowCount.isEmpty) return None
+
+    // save a mutable copy of colStats so that we can later change it recursively
+    mutableColStats = mutable.Map(stats.attributeStats.map(kv => (kv._1.exprId, kv._2)).toSeq: _*)
+
+    // estimate selectivity of this filter predicate
+    val filterSelectivity: Double = calculateConditions(plan.condition)
+
+    // attributeStats has mapping Attribute-to-ColumnStat.
+    // mutableColStats has mapping ExprId-to-ColumnStat.
+    // We use an ExprId-to-Attribute map to facilitate the mapping Attribute-to-ColumnStat
+    val expridToAttrMap: Map[ExprId, Attribute] =
+      stats.attributeStats.map(kv => (kv._1.exprId, kv._1))
+    // copy mutableColStats contents to an immutable AttributeMap.
+    val mutableAttributeStats: mutable.Map[Attribute, ColumnStat] =
+      mutableColStats.map(kv => expridToAttrMap(kv._1) -> kv._2)
+    val newColStats = AttributeMap(mutableAttributeStats.toSeq)
+
+    val filteredRowCount: BigInt =
+      EstimationUtils.ceil(BigDecimal(stats.rowCount.get) * filterSelectivity)
+    val filteredSizeInBytes: BigInt = EstimationUtils.ceil(BigDecimal(
+        EstimationUtils.getOutputSize(plan.output, filteredRowCount, newColStats)
+    ))
+
+    Some(stats.copy(sizeInBytes = filteredSizeInBytes, rowCount = Some(filteredRowCount),
+      attributeStats = newColStats))
+  }
+
+  /**
+   * Returns a percentage of rows meeting a compound condition in Filter node.
+   * A compound condition is decomposed into multiple single conditions linked with AND, OR, NOT.
+   * For logical AND conditions, we need to update stats after a condition estimation
+   * so that the stats will be more accurate for subsequent estimation.  This is needed for
+   * range condition such as (c > 40 AND c <= 50)
+   * For logical OR conditions, we do not update stats after a condition estimation.
+   *
+   * @param condition the compound logical expression
+   * @param update a boolean flag to specify if we need to update ColumnStat of a column
+   *               for subsequent conditions
+   * @return a double value to show the percentage of rows meeting a given condition.
+   *         Returns 1.0 (a conservative filter estimate) if a condition is not supported.
+   */
+  def calculateConditions(condition: Expression, update: Boolean = true): Double = {
+
+    condition match {
+      case And(cond1, cond2) =>
+        val p1 = calculateConditions(cond1, update)
+        val p2 = calculateConditions(cond2, update)
+        p1 * p2
+
+      case Or(cond1, cond2) =>
+        val p1 = calculateConditions(cond1, update = false)
+        val p2 = calculateConditions(cond2, update = false)
+        math.min(1.0, p1 + p2 - (p1 * p2))
+
+      case Not(cond) => calculateSingleCondition(cond, update = false) match {

```

Comment:
Estimation is based on some simplification assumptions. Actually we don't know if we are under-estimating or over-estimating, e.g. estimating c < value by (max-value) / (max-min) based on uniform assumption, it might be under-estimated or over-estimated, depending on actual distribution.
So I think we can support nested Not.